pubsub/pubsub.go

```diff
@@ -427,67 +476,70 @@ func (s *Subscription) Receive(ctx context.Context) (_ *Message, err error) {
 			// The Subscription is in a permanent error state. Return the error.
 			return nil, s.err // s.err wrapped when set
 		}
+
+		if s.waitc == nil && float64(len(s.q)) <= s.runningBatchSize*prefetchRatio {
+			// We think we're going to run out of messages in expectedReceiveBatchDuration,
+			// and there's no outstanding ReceiveBatch call, so initiate one in the
+			// background.
+			// Completion will be signalled to this goroutine, and to any other
+			// waiting goroutines, by closing s.waitc.
+			s.waitc = make(chan struct{})
+			batchSize := s.updateBatchSize()
+
+			go func() {
+				if s.preReceiveBatchHook != nil {
+					s.preReceiveBatchHook(batchSize)
+				}
+				msgs, err := s.getNextBatch(batchSize)
+				if s.postReceiveBatchHook != nil {
+					s.postReceiveBatchHook(len(msgs))
+				}
+				s.mu.Lock()
+				defer s.mu.Unlock()
+				if err != nil {
+					// Non-retryable error from ReceiveBatch -> permanent error.
+					s.err = err
+				} else if len(msgs) > 0 {
+					s.q = append(s.q, msgs...)
+					if s.throughputStart.IsZero() {
+						s.throughputStart = time.Now()
```

Comment:
I noticed I was missing a case in `updateBatchSize` (it should set `throughputStart` to the zero time if there aren't any messages in the queue). Added. Maybe that was confusing you.

The idea is that we want to measure the time between `updateBatchSize` calls when messages are available to be received; we don't want to include time when the queue is empty. Therefore, in `updateBatchSize`, if there are  still messages available, we set `throughputStart` to `now`. If there aren't messages available, we set it to the zero time; it will likely get set to a non-zero value here, when messages arrive (although, it's possible 0 messages arrive and so `throughputStart` is still 0 the next time we get into `updateBatchSize`, in which case we skip the updating part.